exporter/prometheusexporter/exporter.go

```diff
@@ -0,0 +1,350 @@
+	if metric == nil || len(metric.Timeseries) == 0 {
+		return nil
+	}
+
+	// TODO: (@odeke-em) use node, resource and metrics e.g. perhaps to derive default labels
+	return exp.collector.addMetric(metric)
+}
+
+func (c *collector) addMetric(metric *metricspb.Metric) error {
+	if err := c.registerMetrics(metric); err != nil {
+		return err
+	}
+
+	signature := metricSignature(c.opts.Namespace, metric)
+
+	c.mu.Lock()
+	c.metricsData[signature] = metric
+	c.mu.Unlock()
+
+	return nil
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	c.registeredMetricsMu.Lock()
+	registered := make(map[string]*prometheus.Desc)
+	for key, desc := range c.registeredMetrics {
+		registered[key] = desc
+	}
+	c.registeredMetricsMu.Unlock()
+
+	for _, desc := range registered {
+		ch <- desc
+	}
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	// We need a copy of all the metric data up until this point in time.
+	metricsData := c.cloneMetricsData()
+
+	ctx, span := trace.StartSpan(context.Background(), "prometheus.Metrics.Collect", trace.WithSampler(trace.NeverSample()))
+	defer span.End()
+
+	for _, metric := range metricsData {
+		for _, timeseries := range metric.Timeseries {
+			pmetrics, err := c.protoTimeSeriesToPrometheusMetrics(ctx, metric, timeseries)
+			if err == nil {
+				for _, pmetric := range pmetrics {
+					ch <- pmetric
+				}
+			} else {
+				c.opts.onError(err)
+			}
+		}
+	}
+}
+
+var errNilTimeSeries = errors.New("expecting a non-nil TimeSeries")
+
+func (c *collector) protoTimeSeriesToPrometheusMetrics(ctx context.Context, metric *metricspb.Metric, ts *metricspb.TimeSeries) ([]prometheus.Metric, error) {
+	if ts == nil {
+		return nil, errNilTimeSeries
+	}
+
+	labelKeys := metric.GetMetricDescriptor().GetLabelKeys()
+	labelValues, err := protoLabelValuesToLabelValues(labelKeys, ts.LabelValues)
+	if err != nil {
+		return nil, err
+	}
+	derivedPrometheusValueType := prometheusValueType(metric)
+	desc, _, _ := c.lookupPrometheusDesc(c.opts.Namespace, metric)
+
+	pmetrics := make([]prometheus.Metric, 0, len(ts.Points))
+	for _, point := range ts.Points {
+		pmet, err := protoMetricToPrometheusMetric(ctx, point, desc, derivedPrometheusValueType, labelValues)
+		if err == nil {
+			pmetrics = append(pmetrics, pmet)
+		} else {
+			// TODO: (@odeke-em) record these errors
+		}
+	}
+	return pmetrics, nil
+}
+
+func protoLabelValuesToLabelValues(rubricLabelKeys []*metricspb.LabelKey, protoLabelValues []*metricspb.LabelValue) ([]string, error) {
+	if len(protoLabelValues) == 0 {
+		return nil, nil
+	}
+
+	if len(protoLabelValues) > len(rubricLabelKeys) {
+		return nil, fmt.Errorf("len(LabelValues)=%d > len(labelKeys)=%d", len(protoLabelValues), len(rubricLabelKeys))
+	}
+	plainLabelValues := make([]string, len(rubricLabelKeys))
+	for i := 0; i < len(rubricLabelKeys); i++ {
+		if i >= len(protoLabelValues) {
```

Comment:
at this point`len(protoLabelValues) <= len(rubricaLabelKeys)` using that it seem better to limit the loop by `len(protoLabelValues)` and remove this check (or since `i` is going to keep increasing break on the first one).

Reply:
Great idea. I'll fix that.